src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/SourceGenJsonTypeInfoOfT.cs

```diff
@@ -107,8 +107,13 @@ internal override JsonParameterInfoValues[] GetParameterInfoValues()
             JsonParameterInfoValues[] array;
             if (CtorParamInitFunc == null || (array = CtorParamInitFunc()) == null)
             {
-                ThrowHelper.ThrowInvalidOperationException_NoMetadataForTypeCtorParams(Options.TypeInfoResolver, Type);
-                return null!;
+                if (SerializeHandler == null)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_NoMetadataForTypeCtorParams(Options.TypeInfoResolver, Type);
+                }
```

Comment:
wouldn't we get here only if source gen did something wrong? this check makes sense if we plan to make SerializeHandler public but otherwise I'd stick with always throwing

Reply:
This is the root cause of the bug. Essentially, calling `EnsureConfigured()` on a fast-path only JsonTypeInfo would result in it throwing here. This PR changes it so that any exceptions are delayed until serialization/deserialization occurs (via a check on `MetadataSerializationNotSupported`)